crates/astria-composer/src/searcher/mod.rs

```diff
@@ -0,0 +1,174 @@
+    types::Transaction,
+};
+use tokio::{select, task::JoinSet};
+use tracing::warn;
+
+use crate::Config;
+
+#[cfg(test)]
+mod tests;
+
+pub struct Searcher<T>
+where
+    T: PubsubClient + JsonRpcClient + Clone,
+{
+    // The client for getting new pending transactions from the ethereum JSON RPC.
+    eth_client: Provider<T>,
+    // The client for submitting swrapped pending eth transactions to the astria sequencer.
+    sequencer_client: astria_sequencer_client::Client,
+    rollup_chain_id: String,
+    // Set of currently running jobs converting pending eth transactions to signed sequencer
+    // transactions.
+    conversion_tasks: JoinSet<eyre::Result<SignedSequencerTx>>,
+    // Set of in-flight RPCs submitting signed transactions to the sequencer.
+    // submission_tasks: JoinSet<eyre::Result<tx_sync::Response>>,
+    submission_tasks: JoinSet<eyre::Result<()>>,
+}
+
+impl<T> Searcher<T>
+where
+    T: PubsubClient + JsonRpcClient + Clone,
+{
+    pub fn build(
+        eth_client: Provider<T>,
+        sequencer_client: astria_sequencer_client::Client,
+        rollup_chain_id: String,
+    ) -> Searcher<T> {
+        Searcher {
+            eth_client,
+            sequencer_client,
+            rollup_chain_id,
+            conversion_tasks: JoinSet::new(),
+            submission_tasks: JoinSet::new(),
+        }
+    }
+
+    /// Constructs a new Searcher service from config.
+    pub async fn new_ws(cfg: &Config) -> eyre::Result<Searcher<Ws>> {
+        // connect to eth node
+        let eth_client = Provider::connect(&cfg.execution_ws_url)
+            .await
+            .wrap_err("failed connecting to ethereum json rpc server")?;
+
+        // connect to sequencer node
+        let sequencer_client = SequencerClient::new(&cfg.sequencer_url)
+            .wrap_err("failed constructing sequencer client")?;
+
+        let rollup_chain_id = cfg.chain_id.clone();
+
+        Ok(Searcher {
+            eth_client,
+            sequencer_client,
+            rollup_chain_id,
+            conversion_tasks: JoinSet::new(),
+            submission_tasks: JoinSet::new(),
+        })
+    }
+
+    /// Serializes and signs a sequencer tx from a rollup tx.
+    async fn handle_pending_tx(&mut self, rollup_tx: Transaction) -> eyre::Result<()> {
+        let chain_id = self.rollup_chain_id.clone();
+
+        self.conversion_tasks.spawn_blocking(move || {
+            // FIXME: Needs to be altered when nonces are implemented in the sequencer
+            // For now, each transaction is transmitted from a new account with nonce 0
+            let sequencer_key = SigningKey::new(rand::thread_rng());
+            let nonce = Nonce::from(0);
+
+            // Pack into sequencer tx
+            let data = rollup_tx.rlp().to_vec();
+            let chain_id = chain_id.into_bytes();
+            let seq_action = SequencerAction::SequenceAction(SequenceAction::new(chain_id, data));
```

Comment:
Naming here is kinda confusing, perhaps we should import `astria_sequencer::transaction::action::Action`  `TransactionAction` instead of `SequencerTransaction`?

Reply:
I fully agree with this. However, I am going to leave this for now because the actual fix is in creating a proper public interface (via `sequencer-rpc`, for example).